### PR TITLE
Async

### PR DESCRIPTION
**Pull request by kwalrath**
Conversion from Future/Stream APIs to async/await/await for

### CODE REVIEW

File: cmdline/bin/dcat.dart [hunk 1]
```diff
@@ -36,34 +36,35 @@ void main(List<String> arguments) {
   dcat(paths, argResults[LINE_NUMBER]);
 }
 
-Future dcat(List<String> paths, bool showLineNumbers) {
+Future dcat(List<String> paths, bool showLineNumbers) async {
```
Comment (sethladd) [inline cmdline/bin/dcat.dart]: do we need to return Future and mark our function as async ?

Comment (kwalrath) [inline cmdline/bin/dcat.dart]: I'm not sure what you mean. (Btw, this isn't ready for review yet. It's just up there so people like shams can see what I've done, so far.)

Comment (sethladd) [inline cmdline/bin/dcat.dart]: it's sufficient to do:

   dcat(List<String> paths, bool showLineNumbers) async {

Without specifying the Future return type.

Not sure what the right convention is.

Comment (kwalrath) [inline cmdline/bin/dcat.dart]: Ah, right. We've been specifying types for functions but not (as a rule) for the code within them. This is kind of sort of following the [PREFER providing type annotations on public APIs](https://www.dartlang.org/articles/style-guide/#type-annotations) guideline, except for the fact that we're not really defining a public API here. I suppose it makes sense to drop the Future if you're not explicitly returning anything. (On the other hand, it does let people know that there really is a value returned. I'm not sure what to do. @munificent?)

Comment (sethladd) [inline cmdline/bin/dcat.dart]: Our tools can see `async` and infer that the function will return a Future. I think our style guide can ask our developers not to include Future when they already included async in the function/method signature.

Comment (kwalrath) [inline cmdline/bin/dcat.dart]: After further discussion, I believe we've decided that async is an implementation detail that doesn't affect the method signature, so you *should* specify the return value.

Comment (sethladd) [inline cmdline/bin/dcat.dart]: Yup! Use both `Future` and `async`

**Comment by kwalrath** [comment]
@larsbak and @munificent could you please check my diffs and see if there's anything you think should be done differently?

File: cmdline/bin/dcat.dart [hunk 2]
```diff
@@ -36,34 +36,36 @@ void main(List<String> arguments) {
...
     // No files provided as arguments. Read from stdin and print each line.
-    return stdin.pipe(stdout);
+    stdin.pipe(stdout);
   } else {
-    return Future.forEach(paths, (path) {
+    for (var path in paths) {
       int lineNumber = 1;
-      Stream<List<int>> stream = new File(path).openRead();
-      return stream
+      Stream stream = new File(path)
```
Comment (munificent) [inline cmdline/bin/dcat.dart]: I'd call this `lines`. I care more about what it's a stream *of* than I do that it's a stream.

File: count_down/web/count_down.dart [hunk 3]
```diff
@@ -4,55 +4,55 @@
...
- 
+
   // Life-cycle methods...
-  
+
   // Called from the VIEW when the element is inserted into the DOM.
-  Future start() {
+  Future start() async {
     if (!idbAvailable) {
       return new Future.error('IndexedDB not supported.');
```
Comment (munificent) [inline count_down/web/count_down.dart]: throw 'IndexedDB not supported.';

File: count_down/web/count_down.dart [hunk 4]
```diff
@@ -4,55 +4,55 @@
...
-    return _store.open().then((_) {
-      _startMilestoneTimer();
-      hazMilestones = notifyPropertyChange(const Symbol('hazMilestones'),
-          hazMilestones, (milestones.length == 0) ? false : true);
 
-    });
+    await _store.open();
+    _startMilestoneTimer();
+    hazMilestones = notifyPropertyChange(
+        const Symbol('hazMilestones'), hazMilestones,
```
Comment (munificent) [inline count_down/web/count_down.dart]: `const Symbol('hazMilestones')` -> `#hazMilestones`

File: count_down/web/count_down.dart [hunk 5]
```diff
@@ -62,53 +62,58 @@ class MilestoneApp extends Observable {
...
-      _store.add(milestoneName, occursOn).then((_) {
+      try {
+        await _store.add(milestoneName, occursOn);
         _startMilestoneTimer();
-        hazMilestones = notifyPropertyChange(const Symbol('hazMilestones'),
-            hazMilestones, (milestones.length == 0) ? false : true);
-      },
-      onError: (e) { print('duplicate key'); } );
+        hazMilestones = notifyPropertyChange(
+            const Symbol('hazMilestones'), hazMilestones,
```
Comment (munificent) [inline count_down/web/count_down.dart]: Ditto.

File: count_down/web/count_down.dart [hunk 6]
```diff
@@ -62,53 +62,58 @@ class MilestoneApp extends Observable {
...
-      hazMilestones = notifyPropertyChange(const Symbol('hazMilestones'),
-          hazMilestones, (milestones.length == 0) ? false : true);
-    });
+
+  Future clear() async {
+    await _store.clear();
+    _stopMilestoneTimer(false);
+    hazMilestones = notifyPropertyChange(
+        const Symbol('hazMilestones'), hazMilestones,
+        (milestones.length == 0) ? false : true);
```
Comment (munificent) [inline count_down/web/count_down.dart]: `(milestones.length == 0) ? false : true` -> `milestones.isNotEmpty`

File: count_down/web/count_down.dart [hunk 7]
```diff
@@ -62,53 +62,58 @@ class MilestoneApp extends Observable {
...
   }
-   
+
   // Timer stuff.
   // Starts the timer if it's off and there are milestones.
   void _startMilestoneTimer() {
     if (timer == null && milestones.length > 0) {
       // The timer goes off every second.
```
Comment (munificent) [inline count_down/web/count_down.dart]: Kind of a redundant comment. :)

File: count_down/web/count_down.dart [hunk 8]
```diff
@@ -62,53 +62,58 @@ class MilestoneApp extends Observable {
...
+        (timer != null &&
+            milestones.where((m) => !m.elapsed).isEmpty)) {
       timer.cancel();
       timer = null;
     }
   }
-  
+
   // Update the display for each milestone.
   void _tick(Timer _) {
```
Comment (munificent) [inline count_down/web/count_down.dart]: Nit: If you're not using the parameter anyway, I wouldn't bother annotating it.

File: count_down/web/milestone.dart [hunk 9]
```diff
@@ -45,148 +46,144 @@ class Milestone extends Observable {
...
     // Format individual pieces of the display string.
     String days = (d == 0) ? '' : '$d days, ';
     String hours = (h == 0) ? '' : '$h hours, ';
     String minutes = (m == 0) ? '' : '$m minutes, ';
     String seconds = '$s seconds';
-    
-    _displayString = '$days $hours $minutes $seconds until ${milestoneName}';
+
+    _displayString =
+        '$days $hours $minutes $seconds until ${milestoneName}';
```
Comment (munificent) [inline count_down/web/milestone.dart]: Don't need `{}` here.

File: count_down/web/milestone.dart [hunk 10]
```diff
@@ -45,148 +46,144 @@ class Milestone extends Observable {
...
-    var cursors = store.openCursor(autoAdvance: true).asBroadcastStream();
-    cursors.listen((cursor) {
+    var cursors =
+        store.openCursor(autoAdvance: true).asBroadcastStream();
+    await for (var cursor in cursors) {
       // Add milestone to the internal list.
       var milestone = new Milestone.fromRaw(cursor.key, cursor.value);
       milestones.add(milestone);
+    }
+    ;
```
Comment (munificent) [inline count_down/web/milestone.dart]: Kill the `;`.

File: count_down/web/tute_countdown.dart [hunk 11]
```diff
@@ -42,19 +44,18 @@ class CountDownComponent extends PolymerElement {
...
-        ($['addbutton'] as ButtonElement).disabled = true;
-        ($['clearbutton'] as ButtonElement).disabled = true;
+    appObject.start().catchError((e) {
+      ($['addbutton'] as ButtonElement).disabled = true;
+      ($['clearbutton'] as ButtonElement).disabled = true;
 
-        errorMsg = e.toString();
-      });
+      errorMsg = e.toString();
+    });
```
Comment (munificent) [inline count_down/web/tute_countdown.dart]: Make this async and use `try` here?

File: futures/bin/futures2.dart [hunk 12]
```diff
@@ -1,12 +1,17 @@
...
 import 'dart:async';
 
-void printDailyNewsDigest() {
+//void printDailyNewsDigest() {
+//  File file = new File("dailyNewsDigest.txt");
+//  Future future = file.readAsString();
+//  future.then((content) {
+//    print(content);
+//  });
+//}
```
Comment (munificent) [inline futures/bin/futures2.dart]: Delete.

File: count_down/web/tute_countdown.dart (hunk 11 above)
Comment (kwalrath) [inline count_down/web/tute_countdown.dart]: Like this?

```
  attached() async {
    super.attached();
    try {
      appObject.start();
    } catch (e) {
      ($['addbutton'] as ButtonElement).disabled = true;
      ($['clearbutton'] as ButtonElement).disabled = true;
      errorMsg = e.toString();
    }
  }
```

Comment (munificent) [inline count_down/web/tute_countdown.dart]: Yup!

File: count_down/web/count_down.dart (hunk 7 above)
Comment (kwalrath) [inline count_down/web/count_down.dart]: I might fix that later, after I know the code better.

**Comment by kwalrath** [comment]
Thanks for the review, @munificent! I still have some more samples to convert, so... more from me later.

File: count_down/web/tute_countdown.dart [hunk 13]
```diff
@@ -42,19 +44,19 @@ class CountDownComponent extends PolymerElement {
...
     super.attached();
-    appObject.start()
-      .catchError((e) {
-        ($['addbutton'] as ButtonElement).disabled = true;
-        ($['clearbutton'] as ButtonElement).disabled = true;
-
-        errorMsg = e.toString();
-      });
+    try {
+      appObject.start(); // Returns a future.
```
Comment (madsager) [inline count_down/web/tute_countdown.dart]: Drive-by comments: you should use 'await appObject.start();' otherwise the try-catch will have no effect.

Comment (kwalrath) [inline count_down/web/tute_countdown.dart]: Thanks, Mads!

**Comment by kwalrath** [comment]
@Sfshaza could you check whether any of these changes are still relevant?

**Comment by Sfshaza** [comment]
I went through the changes in this PR. Most were already incorporated. Some examples have been removed after this PR went live. I also took the opportunity to run all of the examples through dartfmt again.

Closing.

[commit](https://github.com/dart-lang/dart-tutorials-samples/commit/0964150d023a3f5f183c444532b81e70afe45a2c)
